### DISCUSSION TITLE
上课时说的“比较”方法似乎行不通？

### DISCUSSION THREAD

**Discussion by QlDoors** [SutuLabs/zkcourse#9] [Week2] [answered]
课堂上的方法代码实现

```c
pragma circom 2.1.4;

include "circomlib/poseidon.circom";
// include "https://github.com/0xPARC/circom-secp256k1/blob/master/circuits/bigint.circom";

template Num2Bits(n) {
    signal input in;
    signal output out[n];
    var lc1 = 0;
    var e2 = 1;
    for (var i = 0; i < n; i++) {
        out[i] <-- (in >> i) & 1;
        out[i] * (out[i] - 1) === 0;
        lc1 += out[i] * e2;
        e2 = e2 + e2;
    }
    log("lc1", lc1);
    log("in", in);
    lc1 === in;
}

template IsEqual (n) {
    signal input in[2];


    signal inv;
    in[0] + (2 ** n) - in[1] ==> inv;  // 这个数最大应该是 2^n + 2^n - 0 = 2 * 2^n，位数为 n + 2

    var nBits = n + 2;
    component n2b = Num2Bits(nBits);
    n2b.in <== inv;
    signal output out0[nBits];
    n2b.out ==> out0;
    signal output out;
    n2b.out[nBits - 1] ==> out;
}

component main = IsEqual(3);
```

####  输入不等

输入：

```
/* INPUT = {
    "in": ["8", "1"]
} */
```

输出：

```
OUTPUT: 
  out0[0] = 1
  out0[1] = 1
  out0[2] = 1
  out0[3] = 1
  out0[4] = 0
  out = 0
```

输入：

```
/* INPUT = {
    "in": ["4", "1"]
} */
```

输出：

```
OUTPUT: 
  out0[0] = 1
  out0[1] = 1
  out0[2] = 0
  out0[3] = 1
  out0[4] = 0
out = 0
```

这个方法似乎行不通，或者这里的n和需要对比的数之间有个约束我没搞对？

**Comment by readygo67** [comment, marked as answer]
1. IsEqual 这个电路实现的像是GreaterOrEqual的功能.
2. input = 8 超过了 3bits 能表示的范围，最大input 应该为7。 为了处理in[0]=8, in[1]=0的情况，将nbits= n+2, 用来表示8+8-0= 0b10000的情况. 会导致3.
3. 理论上 2^3 借到的8是放在bit[3]上，应该检测相减之后的bit[3]， 而原代码检测 n2b.out[nBits-1](bit[4]), 这个信号的含义不明确。
   in[0]       in[1]         inv(5bits).   inv(4bits)
   8             0             0b10000      无法表示                     //5bits. 取最高位 = 1
   8             1              0b01111        0b1111                       //4bits表示, 最高位=1， 2^ 3=8 没有被借走，所以in[0] >= in[1]
    7             0            0b01111        0b1111                       //4bits表示,   最高位=1,   2^ 3=8 没有被借走，所以in[0] >= in[1]
   4             4             0b01000      0b1000                     //4bits表示，最高位=1,   2^ 3=8 没有被借走，所以in[0] >=in[1]
    2             7            0b00011      0b0011                     //4bits表示，最高位=0,   2^ 3=8 被借走，所以in[0] <in[1]


![image](https://user-images.githubusercontent.com/78890754/235219961-61906a3e-373b-4dba-8f01-4e8eec1df5b5.png)

修改后的代码供参考。
<img width="1416" alt="image" src="https://user-images.githubusercontent.com/78890754/235219828-5a5d85d1-d4f9-4ce8-b4c0-d0a012e4524e.png">

**Reply by QlDoors** [reply]
谢谢解答。
我搞错了，把上课讲的比较大小的电路用做判等了。
